src/engine/BMGame.php

```diff
@@ -1979,36 +1979,66 @@ public function valid_attack_types() {
 
         // find out if there are any possible attacks with any combination of
         // the attacker's and defender's dice
-        foreach ($attackTypeArray as $idx => $attackType) {
-            $this->attack = array('attackerPlayerIdx' => $attackerIdx,
-                                  'defenderPlayerIdx' => $defenderIdx,
-                                  'attackerAttackDieIdxArray' =>
-                                      range(0, count($this->activeDieArrayArray[$attackerIdx]) - 1),
-                                  'defenderAttackDieIdxArray' =>
-                                      range(0, count($this->activeDieArrayArray[$defenderIdx]) - 1),
-                                  'attackType' => $attackTypeArray[$idx]);
-            $attack = BMAttack::get_instance($attackType);
-            foreach ($this->activeDieArrayArray[$attackerIdx] as $attackDie) {
-                $attack->add_die($attackDie);
-            }
-            if ($attack->find_attack($this)) {
+        foreach ($attackTypeArray as $attackType) {
+            if ($this->does_valid_attack_exist($attackerIdx, $defenderIdx, $attackType, FALSE)) {
                 $validAttackTypeArray[$attackType] = $attackType;
             }
         }
 
-        $this->attack = $attackCache;
-
         if (empty($validAttackTypeArray)) {
             $validAttackTypeArray['Pass'] = 'Pass';
+
+            // Now find optional attack types
+
+            // Currently, the only optional attacks are skill attacks
+            // involving Warrior dice, so the code is streamlined to search
+            // specifically for this. However, this could easily be generalised
+            // if other optional attack types become available.
+            $doWarriorDiceExist = FALSE;
+
+            foreach ($this->activeDieArrayArray[$attackerIdx] as $activeDie) {
+                if ($activeDie->has_skill('Warrior')) {
+                    $doWarriorDiceExist = TRUE;
+                    break;
+                }
+            }
+
+            if ($doWarriorDiceExist) {
+                if ($this->does_valid_attack_exist($attackerIdx, $defenderIdx, 'Skill', TRUE)) {
+                    $validAttackTypeArray[$attackType] = $attackType;
```

Comment:
I found a bug and tracked it down to this line of code --- you don't want `$attackType` here, you want `Skill`.  Because you use but don't define `$attackType`, it takes whatever the most recent value was in the iteration above.  In a game with no skills other than Warrior, that may well be Skill.  In a game like this one, the API is telling me that i have a valid Speed attack (and *isn't* telling me i have a valid Skill attack --- it lists my options as Default, Pass, and Speed), when it should obviously be the other way around:

![api_erroneous_speed_attack](https://cloud.githubusercontent.com/assets/2898207/6201526/6d930cda-b47d-11e4-9441-779b2ce089e1.png)